### PR TITLE
[wip] Tx pool filter

### PR DESCRIPTION
**Pull request by aeyakovenko**
#### Problem

TX prioritization is expensive

#### Summary of Changes

This is a transaction pool, it should sit before and after sig verify. It used to prioritize which TXs should be considered for inclusion.  

Perf:

test bench_pool_insert   ... bench:         286 ns/iter (+/- 85)
test bench_pool_pop      ... bench:     504,591 ns/iter (+/- 10,995)

Pop is around 500 txs per batch

1. add all the txs into the pool2. 
2. sort them by fee/cu
3. pull X amount of blocks worth from the pool, at least 8, 4 for the current leader and 4 for the next
4. when pulling do not exceed the bucket limits per writable account
5. compute the min fee required per bucket and cheapest block during 4
6. use the fee from 5 as a pre-sigverify filter.
7. clear the pre-sigverify filters any time its not saturated after 8 blocks worth of txs are taken from the pool.

The pre-sigverify filter is TBD.  But should be easy to build in the loop that pops all the txs.

@taozhu-chicago @buffalu @sakridge @carllin

### CODE REVIEW

File: transaction-pool/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+}
+
+impl Pool {
+    pub fn insert(&mut self, item: Item) {
+        let bucket = self
+            .items
+            .entry(item.priority)
+            .or_insert_with(VecDeque::new);
+        bucket.push_back(item);
+        self.count += 1;
```
Comment (taozhu-chicago) [inline transaction-pool/src/lib.rs:34]: Should `pool` has max capacity, say up to 128,000 packets; Drop low priority packets when `self.count > self.capacity` by `self.first_entry().pop_front()`? When a bucket is empty after pop, maybe OK to keep the empty bucket in tree.

File: transaction-pool/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+    fn keys(&self, item: &Item) -> &[&Pubkey];
+}
+
+impl Pool {
+    pub fn insert(&mut self, item: Item) {
+        let bucket = self
+            .items
+            .entry(item.priority)
+            .or_insert_with(VecDeque::new);
+        bucket.push_back(item);
```
Comment (taozhu-chicago) [inline transaction-pool/src/lib.rs:33]: So items are in time-order if same priority? In normal condition, assume most packets will fall into few buckets, to the extreme where a pool is just one VecDeque?

File: transaction-pool/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+    pub count: usize,
+    items: BTreeMap<u64, VecDeque<Item>>,
+}
+
+pub trait Table {
+    fn keys(&self, item: &Item) -> &[&Pubkey];
+}
+
+impl Pool {
+    pub fn insert(&mut self, item: Item) {
```
Comment (taozhu-chicago) [inline transaction-pool/src/lib.rs:28]: Are you envisioning Vote and Transactions are mixed in same pool?

File: transaction-pool/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+    pub id: (u32, u32),
+}
+
+#[derive(Default, Clone)]
+pub struct Pool {
+    pub max_age: u64,
+    pub max_bucket_cu: u64,
+    pub max_block_cu: u64,
+    pub count: usize,
+    items: BTreeMap<u64, VecDeque<Item>>,
```
Comment (taozhu-chicago) [inline transaction-pool/src/lib.rs:20]: Assuming `Item`s in the Pool are `Packet` not `PacketBatch`? Perhaps `SigVerify` receives and sends `Packet`s?

File: transaction-pool/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+        bucket.push_back(item);
+        self.count += 1;
+    }
+    pub fn pop_block<T: Table>(&mut self, time: u64, table: &T) -> VecDeque<Item> {
+        let mut rv = VecDeque::new();
+        let mut total_cu: u64 = 0;
+        let mut block_full = false;
+        let mut buckets: HashMap<Pubkey, u64> = HashMap::new();
+        let mut gc = vec![];
+        for (k, v) in &mut self.items.iter_mut() {
```
Comment (taozhu-chicago) [inline transaction-pool/src/lib.rs:42]: need revert iter in order to go from high priority to low?

File: transaction-pool/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+impl Pool {
+    pub fn insert(&mut self, item: Item) {
+        let bucket = self
+            .items
+            .entry(item.priority)
+            .or_insert_with(VecDeque::new);
+        bucket.push_back(item);
+        self.count += 1;
+    }
+    pub fn pop_block<T: Table>(&mut self, time: u64, table: &T) -> VecDeque<Item> {
```
Comment (taozhu-chicago) [inline transaction-pool/src/lib.rs:36]: After a block-worth packets are prepared and sent to banking_stage for processing; If some of packets in `VecDeque<Item>` are `retryable`, are they being re-inserted back into pool?

**Comment by taozhu-chicago** [comment]
> sort them by fee/cu

Calculating `fee/cu` requires current working bank, at least a recent bank, in order to get current `fee_structure`, does it? 
 
```
fee/cu 
= total_fee/requested_cu
= [additional_fee + 
    sig_fee( num_of_sig, bank.fee_structure ) + 
    write_lock_fee( num_of_write, bank.fee_structure) + 
    compute_fee( requested_cu, bank.fee_structure) 
    ] / requested_cu
```
Not sure how frequent fee_structure is updated

File: transaction-pool/src/lib.rs (hunk 2 above)
Comment (aeyakovenko) [reply transaction-pool/src/lib.rs:37]: yea, basically, and retry puts it in the back of the fifo queue.  not sure if that is right, or should be up front.  up front would preserve the order.

File: transaction-pool/src/lib.rs (hunk 3 above)
Comment (aeyakovenko) [reply transaction-pool/src/lib.rs:32]: just thinking about non vote txs for now

File: transaction-pool/src/lib.rs (hunk 4 above)
Comment (aeyakovenko) [reply transaction-pool/src/lib.rs:24]: Yea, these are just pointers, u32 id for the PacketBatch and u32 for the index in the packet.

File: transaction-pool/src/lib.rs (hunk 5 above)
Comment (aeyakovenko) [reply transaction-pool/src/lib.rs:47]: or we invert the value.

File: transaction-pool/src/lib.rs (hunk 6 above)
Comment (aeyakovenko) [reply transaction-pool/src/lib.rs:40]: Yea, I thought we would re-insert txs back into the pool

**Comment by aeyakovenko** [comment]
> > sort them by fee/cu
> 
> Calculating `fee/cu` requires current working bank, at least a recent bank, in order to get current `fee_structure`, does it?
> 
> ```
> fee/cu 
> = total_fee/requested_cu
> = [additional_fee + 
>     sig_fee( num_of_sig, bank.fee_structure ) + 
>     write_lock_fee( num_of_write, bank.fee_structure) + 
>     compute_fee( requested_cu, bank.fee_structure) 
>     ] / requested_cu
> ```
> 
> Not sure how frequent fee_structure is updated

that should be deterministic, right @jackcmay ?

File: transaction-pool/src/lib.rs (hunk 5 above)
Comment (aeyakovenko) [reply transaction-pool/src/lib.rs]: rev is easy enough to keep it simple

**Pull request closed without merging**